packages/contracts/contracts/plugin/PluginInstaller.sol

```diff
@@ -0,0 +1,360 @@
+    error PluginCountTooBig();
+    error UpdateNotAllowed();
+    error AlreadyThisVersion();
+    error UpgradeNotExistOnProxy();
+    error NotSupportsUpgradable();
+
+    error PermissionsWrong();
+    error PluginNotDeployed();
+    error HelpersWrong();
+    error PluginRepoNotExists();
+    error PluginWithTheSameAddressExists();
+
+    /// @notice Thrown after the plugin installation to detect plugin was installed on a dao.
+    /// @param dao The dao address that plugin belongs to.
+    /// @param plugin the plugin address.
+    event PluginInstalled(address dao, address plugin);
+
+    event Deployed(
+        address indexed sender,
+        address indexed dao,
+        address indexed plugin,
+        address pluginManager,
+        address[] helpers,
+        Permission.ItemMultiTarget[] permissions,
+        bytes data
+    );
+
+    event Updated(address indexed dao, address[] helpers, Permission.ItemMultiTarget[] permissions);
+    event PluginUpdated(address indexed dao, address indexed plugin);
+    event PluginUninstalled(address indexed dao, address indexed plugin);
+
+    /// @dev Modifier used to check if caller is the DAO, or has given permission by the DAO.
+    /// @param _dao The address of the DAO.
+    modifier daoAuthorized(address _dao) {
+        _isCallAllowed(_dao);
+        _;
+    }
+
+    constructor(AragonPluginRegistry _repoRegistry, address _dao) {
+        __DaoAuthorizable_init(IDAO(_dao));
+        repoRegistry = _repoRegistry;
+    }
+
+    function setRepoRegistry(AragonPluginRegistry _repoRegistry)
+        external
+        auth(SET_REPO_REGISTRY_PERMISSION_ID)
+    {
+        repoRegistry = _repoRegistry;
+    }
+
+    function deployInstall(
+        address dao,
+        address pluginManager,
+        PluginRepo pluginManagerRepo,
+        bytes memory data // encoded per pluginManager's deploy ABI
+    ) public returns (Permission.ItemMultiTarget[] memory) {
+        // ensure repo for plugin manager exists
+        if (!repoRegistry.entries(address(pluginManagerRepo))) {
+            revert PluginRepoNotExists();
+        }
+
+        // Reverts if pluginManager doesn't exist on the repo...
+        pluginManagerRepo.getVersionByPluginManager(pluginManager);
+
+        // deploy
+        (
+            address plugin,
+            address[] memory helpers,
+            Permission.ItemMultiTarget[] memory permissions
+        ) = PluginManager(pluginManager).deploy(dao, data);
+
+        // important safety measure to include dao + plugin manager in the encoding.
+        bytes32 hash = keccak256(abi.encode(dao, pluginManager, plugin));
+
+        pluginInstalledChecker[hash] = true;
+
+        installPermissionHashes[hash] = getPermissionsHash(permissions);
+
+        helperHashes[hash] = keccak256(abi.encode(helpers));
+
+        emit Deployed(msg.sender, dao, plugin, pluginManager, helpers, permissions, data);
+
+        return permissions;
+    }
+
+    function installPlugin(
```

Comment:
The function names don't tell me the difference between `deployInstall` and `installPlugin`

Reply:
Suggestion: `registerPluginInstallation` vs `applyPluginInstallation`